File: comm/interfaces_src/IgluMessage.java
package interfaces;

/**
 * <code>IgluMessage</code> class used to format messages between processes
 * Programmer: Christian Wagner
 * Date Created: 1/4/2020
 * Date Modified: 11/02/2020
 * Version: 2.0
 */
public class IgluMessage {
    // defaults
    private final String DEFAULT_MSG = "Message not set"; // default message text
    private final String DEFAULT_TIMESTAMP = "0000-00-00 00:00:00"; // default timestamp text
    private final String ERROR_MSG = "Error building message object"; // error message text

    // message data delimiter setting
    private final String FORMAT_DELIMITER = "&";

    // object variables
    private String message; // message text
    private String timestamp; // message time on sending timestamp
    private int priority; // message priority level
    private int target; // message target Process ID (PID)
    private int sender; // message sender's PID for return messages
    private boolean encoded; // message encrypted flag

    /**
     * Default constructor for the <code>IgluMessage</code> class
     */
    public IgluMessage() {
        this.message = DEFAULT_MSG;
        this.timestamp = DEFAULT_TIMESTAMP;
        this.priority = 0;
        this.target = 0;
        this.sender = 0;
        this.encoded = false;
    }

    /**
     * Overloaded constructor for the <code>IgluMessage</code> class
     * Constructs a <code>IgluMessage</code> object using given data
     * @param message message text
     * @param timestamp message time on sending timestamp
     * @param priority message priority level
     * @param target message target Process ID (PID)
     * @param sender message sender's PID for return messages
     */
    public IgluMessage(String message, String timestamp, int priority, int target, int sender) {
        this.message = message;
        this.timestamp = timestamp;
        this.priority = priority;
        this.target = target;
        this.sender = sender;
        this.encoded = false;
    }

    /**
     * Overloaded constructor for <code>IgluMessage</code> class
     * Constructs a <code>IgluMessage</code> object using raw data from the server
     * @param rawData data stream from server
     */
    public IgluMessage(String rawData) {
        String[] pieces = rawData.split(FORMAT_DELIMITER); // parse raw data

        // check length
        if(pieces.length == 6) {
            // get message part
            String[] tmp = pieces[0].split("=");
            if(tmp[0].contains("MESSAGE") && tmp.length == 2) this.message = tmp[1];
            else this.message = ERROR_MSG;

            // get timestamp part
            tmp = pieces[1].split("=");
            if(tmp[0].contains("TIMESTAMP") && tmp.length == 2) this.timestamp = tmp[1];
            else this.timestamp = DEFAULT_TIMESTAMP;

            // get priority part
            tmp = pieces[2].split("=");
            if(tmp[0].contains("PRIORITY") && tmp.length == 2) this.priority = Integer.parseInt(tmp[1]);
            else this.priority = 0;

            // get target part
            tmp = pieces[3].split("=");
            if(tmp[0].contains("TARGET") && tmp.length == 2) this.target = Integer.parseInt(tmp[1]);
            else this.target = 0;

            // get sender part
            tmp = pieces[4].split("=");
            if(tmp[0].contains("SENDER") && tmp.length == 2) this.sender = Integer.parseInt(tmp[1]);
            else this.sender = 0;

            // get encoded part
            tmp = pieces[5].split("=");
            if(tmp[0].contains("ENCODED") && tmp.length == 2) this.encoded = (tmp[1].equals("1") ? true : false);
            else this.encoded = false;
        } else {
            this.message = ERROR_MSG;
            this.timestamp = DEFAULT_TIMESTAMP;
            this.priority = 0;
            this.target = 0;
            this.sender = 0;
            this.encoded = false;
        }
    }

    /**
     * Returns the message text if the message is not encoded
     * @return the message text if encoded is false, "encoded" otherwise
     */
    public String getMessage() {
        if(!this.encoded) return this.message;
        else return "encoded";
    }

    /**
     * Returns timestamp text
     * @return timestamp text
     */
    public String getTimestamp() {
        return this.timestamp;
    }

    /**
     * Returns priority level
     * @return priority level
     */
    public int getPriority() {
        return this.priority;
    }

    /**
     * Returns target Process ID (PID)
     * @return target process ID (PID)
     */
    public int getTarget() {
        return this.target;
    }

    /**
     * Returns sender's Process ID (PID)
     * @return sender's process ID (PID)
     */
    public int getSender() {
        return this.sender;
    }

    /**
     * Returns value of encoded flag
     * @return 1 if encoded, 0 otherwise
     */
    public int isEncoded() {
        if(encoded) return 1;
        else return 0;
    }

    /**
     * Gets text form of data stored in <code>IgluMessage</code> object for sending to the server
     * @return the text form of the <code>IgluMessage</code> object
     */
    public String getTxData() {
        String sendData = "MESSAGE=" + this.message;
        sendData += FORMAT_DELIMITER + "TIMESTAMP=" + this.timestamp;
        sendData += FORMAT_DELIMITER + "PRIORITY=" + this.priority;
        sendData += FORMAT_DELIMITER + "TARGET=" + this.target;
        sendData += FORMAT_DELIMITER + "SENDER=" + this.target;
        sendData += FORMAT_DELIMITER + "ENCODED=" + ((this.encoded) ? 1 : 0);
        return sendData;
    }

    /**
     * Encrypts the message with a specific encoding and key
     * @param encoding the encoding to use
     * @param key the key to encrypt with
     */
    public void encode(String encoding, String key) {this.encoded = true;}

    /**
     * Decrypts the message with a specific encoding and key
     * @param encoding the encoding to use
     * @param key the key to encrypt with
     */
    public void decode(String encoding, String key) {this.encoded = false;}
}

File: comm/interfaces_src/IgluMessage.py
# IgluMessage class used to format messages between processes
# Programmer: Christian Wagner
# Date Created: 1/09/2020
# Date Modified: 11/02/2020
# Version: 2.0

class IgluMessage:
    # constructor for raw data from the server
    def __init__(self, *args):
        self.__format_delimiter = '&' # message data delimiter setting
        if len(args) == 5:
            # new message without raw data
            self.__message = args[0]
            self.__timestamp = args[1]
            self.__priority = args[2]
            self.__target = args[3]
            self.__sender = args[4]
            self.__encoded = False
        elif len(args) == 1:
            # new message with raw data
            pieces = args[0].split(self.__format_delimiter) # parse raw data
            # check length
            if len(pieces) == 6:
                tmp = pieces[0].split("=")
                if 'MESSAGE' in tmp[0] and len(tmp) == 2:
                    self.__message = tmp[1]
                else:
                    self.__message = 'Error building message object'

                tmp = pieces[1].split("=")
                if 'TIMESTAMP' in tmp[0] and len(tmp) == 2:
                    self.__timestamp = tmp[1]
                else:
                    self.__timestamp = '0000-00-00 00:00:00'
            
                tmp = pieces[2].split("=")
                if 'PRIORITY' in tmp[0] and len(tmp) == 2:
                    self.__priority = tmp[1]
                else:
                    self.__priority = 0

                tmp = pieces[3].split("=")
                if 'TARGET' in tmp[0] and len(tmp) == 2:
                    self.__target = tmp[1]
                else:
                    self.__target = 0

                tmp = pieces[4].split("=")
                if 'SENDER' in tmp[0] and len(tmp) == 2:
                    self.__sender = tmp[1]
                else:
                    self.__sender = 0

                tmp = pieces[5].split("=")
                if 'ENCODED' in tmp[0] and len(tmp) == 2:
                    self.__encoded = False if tmp[1] == '0' else True
                else:
                    self.__encoded = False
            else:
                self.__message = 'Error building message object'
                self.__timestamp = '0000-00-00 00:00:00'
                self.__priority = 0
                self.__target = 0
                self.__sender = 0
                self.__encoded = False
        else:
            self.__message = 'Message not set'
            self.__timestamp = '0000-00-00 00:00:00'
            self.__priority = 0
            self.__target = 0
            self.__sender = 0
            self.__encoded = False

    # getter for __message variable
    @property
    def message(self):
        return self.__message

    # getter for __timestamp variable
    @property
    def timestamp(self):
        return self.__timestamp

    # getter for __priority variable
    @property
    def priority(self):
        return self.__priority

    # getter for __target variable
    @property
    def target(self):
        return self.__target

    # getter for __sender variable
    @property
    def sender(self):
        return self.__sender

    # getter for __encoded variable
    @property
    def encoded(self):
        return self.__encoded

    # Returns text form of data stored in IgluMessage object for sending to the server
    def get_tx_data(self):
        tmp = 'MESSAGE=' + self.__message
        tmp += self.__format_delimiter + 'TIMESTAMP=' + self.__timestamp
        tmp += self.__format_delimiter + 'PRIORITY=' + str(self.__priority)
        tmp += self.__format_delimiter + 'TARGET=' + str(self.__target)
        tmp += self.__format_delimiter + 'SENDER=' + str(self.__sender)
        tmp += self.__format_delimiter + 'ENCODED=' + ('1' if self.__encoded else '0')
        return tmp

    # Encrypts the message with a specific encoding and key
    # encoding -> the encoding to use
    # key -> the key to encrypt with
    def encode(self,encoding,key):
        self.__encoded = True

    # Decrypts the message with a specific encoding and key
    # encoding -> the encoding to use
    # key -> the key to encrypt with
    def decode(self,encoding,key):
        self.__encoded = False